[PT Run: Settings Plugin] "System.ArgumentNullException" on queries with `$a`

### Microsoft PowerToys version

0.1.0

### Running as admin

- [ ] Yes

### Area(s) with issue?

PowerToys Run

### Steps to reproduce

Typing a query like `$a`.

### ✔️ Expected Behavior

Getting results.

### ❌ Actual Behavior

Found this exception in log. (Note: `$` is the plugin key word.)

```
[2021-09-10 17:59:46.8172] [ERROR] [D:\AppDev\Github\htcfreek\PowerToys_SettingsPluginPart3\src\modules\launcher\PowerLauncher\Plugin\PluginManager.cs::201]
-------------------------- Begin exception --------------------------
Message: Exception for plugin <Windows Settings> when query <$a>

Exception full name  : System.ArgumentNullException
Exception message    : Value cannot be null. (Parameter 'source')
Exception stack trace:
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper.ResultHelper.SetScores(IEnumerable`1 resultList, String query) in D:\AppDev\Github\htcfreek\PowerToys_SettingsPluginPart3\src\modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.WindowsSettings\Helper\ResultHelper.cs:line 170
   at Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper.ResultHelper.GetResultList(IEnumerable`1& list, String query, String& iconPath) in D:\AppDev\Github\htcfreek\PowerToys_SettingsPluginPart3\src\modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.WindowsSettings\Helper\ResultHelper.cs:line 50
   at Microsoft.PowerToys.Run.Plugin.WindowsSettings.Main.Query(Query query) in D:\AppDev\Github\htcfreek\PowerToys_SettingsPluginPart3\src\modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.WindowsSettings\Main.cs:line 108
   at PowerLauncher.Plugin.PluginManager.<>c__DisplayClass23_1.<QueryForPlugin>b__0() in D:\AppDev\Github\htcfreek\PowerToys_SettingsPluginPart3\src\modules\launcher\PowerLauncher\Plugin\PluginManager.cs:line 179
   at Wox.Infrastructure.Stopwatch.Debug(String message, Action action) in D:\AppDev\Github\htcfreek\PowerToys_SettingsPluginPart3\src\modules\launcher\Wox.Infrastructure\Stopwatch.cs:line 29
   at PowerLauncher.Plugin.PluginManager.QueryForPlugin(PluginPair pair, Query query, Boolean delayedExecution) in D:\AppDev\Github\htcfreek\PowerToys_SettingsPluginPart3\src\modules\launcher\PowerLauncher\Plugin\PluginManager.cs:line 171
Exception source     : System.Linq
Exception target site: Void ThrowArgumentNullException(System.Linq.ExceptionArgument)
Exception HResult    : -2147467261
-------------------------- End exception --------------------------
```

### Other Software

_No response_

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
@@ -188,5 +188,52 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
                 result.Score = lowScore--;
             }
         }
+
+        /// <summary>
+        /// Checks if a setting <see cref="WindowsSetting"/> matches the search string <see cref="Query.Search"/> to filter settings by settings path.
+        /// This method is called from the filter function <see cref="Main.Query(Query)"/> if the search string <see cref="Query.Search"/> contains the character ">".
+        /// </summary>
+        /// <param name="found">The WindowsSetting taht should be checked.</param>
+        /// <param name="searchString">The searchString entered by the user <see cref="Query.Search"/>s.</param>
+        internal static bool FilterBySettingsPath(in WindowsSetting found, in string searchString)
+        {
+            if (!searchString.Contains('>'))
+            {
+                return false;
+            }
+
+            var pathElements = searchString.Split('>');
+
+            // check app name of the WindowsSetting
+            if (!found.Type.StartsWith(pathElements[0]))
+            {
+                return false;
+            }
+
+            // Check area path of the WindowsSetting.
+            // If "pathElements.Length" is higher than one the "searchString" contains at least one area name.
+            // Because the area names in pathElements array start at index 1 sometimes <pathElements.Length - 1> or <index - 1> is used.
+            if (pathElements.Length > 1 & !(found.Areas is null))
+            {
+                if ((pathElements.Length - 1) > found.Areas.Count())
+                {
+                    // The user entered more area names (pathElements[1] -> pathElements[n]) than defined for this WindowsSetting.
+                    return false;
+                }
+
+                for (int i = 1; i <= pathElements.Length; i++)
+                {
+#pragma warning disable CS8602 // Do not catch exception "Possible null reference". The compiler thinks that Areas could be null which is not possible because of the parental if condition.
+                    if (!found.Areas[i - 1].StartsWith(pathElements[i]))
+                    {
+                        // The user has entered an area name that not matches.
+                        return false;
+                    }
+                }
+            }
+
+            // Return "true" if WindowsSetting "WindowsSetting" matches "searchString".
+            return true;
+        }
     }
 }

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
@@ -167,6 +167,12 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
                     continue;
                 }
 
+                if (windowsSetting.Areas is null)
+                {
+                    result.Score = lowScore--;
+                    continue;
+                }
+
                 if (windowsSetting.Areas.Any(x => x.StartsWith(query, StringComparison.CurrentCultureIgnoreCase)))
                 {
                     result.Score = lowScore--;

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
@@ -226,7 +226,7 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
             {
                 if (string.IsNullOrWhiteSpace(queryElements[i]))
                 {
-                    // The queryEelent is an WhiteSpace. Nothing to compare.
+                    // The queryElement is an WhiteSpace. Nothing to compare.
                     break;
                 }
 

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
@@ -154,41 +154,44 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
                     continue;
                 }
 
-                if (windowsSetting.Name.StartsWith(query, StringComparison.CurrentCultureIgnoreCase))
+                if (!string.IsNullOrWhiteSpace(query))
                 {
-                    result.Score = highScore--;
-                    continue;
-                }
+                    if (windowsSetting.Name.StartsWith(query, StringComparison.CurrentCultureIgnoreCase))
+                    {
+                        result.Score = highScore--;
+                        continue;
+                    }
 
-                // If query starts with second or next word of name, set score.
-                if (windowsSetting.Name.Contains($" {query}", StringComparison.CurrentCultureIgnoreCase))
-                {
-                    result.Score = mediumScore--;
-                    continue;
-                }
+                    // If query starts with second or next word of name, set score.
+                    if (windowsSetting.Name.Contains($" {query}", StringComparison.CurrentCultureIgnoreCase))
+                    {
+                        result.Score = mediumScore--;
+                        continue;
+                    }
 
-                if (windowsSetting.Areas is null)
-                {
-                    result.Score = lowScore--;
-                    continue;
-                }
+                    if (windowsSetting.Areas is null)
+                    {
+                        result.Score = lowScore--;
+                        continue;
+                    }
 
-                if (windowsSetting.Areas.Any(x => x.StartsWith(query, StringComparison.CurrentCultureIgnoreCase)))
-                {
-                    result.Score = lowScore--;
-                    continue;
-                }
+                    if (windowsSetting.Areas.Any(x => x.StartsWith(query, StringComparison.CurrentCultureIgnoreCase)))
+                    {
+                        result.Score = lowScore--;
+                        continue;
+                    }
 
-                if (windowsSetting.AltNames is null)
-                {
-                    result.Score = lowScore--;
-                    continue;
-                }
+                    if (windowsSetting.AltNames is null)
+                    {
+                        result.Score = lowScore--;
+                        continue;
+                    }
 
-                if (windowsSetting.AltNames.Any(x => x.StartsWith(query, StringComparison.CurrentCultureIgnoreCase)))
-                {
-                    result.Score = mediumScore--;
-                    continue;
+                    if (windowsSetting.AltNames.Any(x => x.StartsWith(query, StringComparison.CurrentCultureIgnoreCase)))
+                    {
+                        result.Score = mediumScore--;
+                        continue;
+                    }
                 }
 
                 result.Score = lowScore--;

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
@@ -199,7 +199,7 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
         /// Checks if a setting <see cref="WindowsSetting"/> matches the search string <see cref="Query.Search"/> to filter settings by settings path.
         /// This method is called from the <see cref="Predicate{T}"/> method in <see cref="Main.Query(Query)"/> if the search string <see cref="Query.Search"/> contains the character ">".
         /// </summary>
-        /// <param name="found">The WindowsSetting taht should be checked.</param>
+        /// <param name="found">The WindowsSetting's result that should be checked.</param>
         /// <param name="queryString">The searchString entered by the user <see cref="Query.Search"/>s.</param>
         internal static bool FilterBySettingsPath(in WindowsSetting found, in string queryString)
         {
@@ -217,26 +217,24 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
             }
 
             // Check area path of the WindowsSetting.
-            // If "pathElements.Length" is higher than one the "searchString" contains at least one area name.
-            // Because the area names in pathElements array start at index 1 sometimes <pathElements.Length - 1> or <index - 1> is used.
             if (!string.IsNullOrEmpty(queryElements[1]))
             {
                 if (found.Areas is null)
                 {
-                    // The user entered at least one area name (pathElements[1] -> pathElements[n]). But the WindowsSetting nas no defined areas.
+                    // The user entered at least one area name (queryElements[1]). But the WindowsSetting has no defined areas.
                     return false;
                 }
 
                 if ((queryElements.Length - 1) > found.Areas.Count())
                 {
-                    // The user entered more area names (pathElements[1] -> pathElements[n]) than defined for this WindowsSetting.
+                    // The user entered more area names (queryElements[1] -> queryElements[n]) than defined for this WindowsSetting.
                     return false;
                 }
 
                 int areaCounter = 0;
-                foreach (var qEelement in queryElements.Skip(1))
+                foreach (var qElement in queryElements.Skip(1))
                 {
-                    if (!found.Areas[areaCounter].StartsWith(qEelement, StringComparison.CurrentCultureIgnoreCase))
+                    if (!found.Areas[areaCounter].StartsWith(qElement, StringComparison.CurrentCultureIgnoreCase))
                     {
                         // The user has entered an area name that not matches.
                         return false;
@@ -246,7 +244,7 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
                 }
             }
 
-            // Return "true" if WindowsSetting "WindowsSetting" matches "searchString".
+            // Return "true" if <found> matches <queryString>.
             return true;
         }
     }

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
@@ -197,21 +197,21 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
 
         /// <summary>
         /// Checks if a setting <see cref="WindowsSetting"/> matches the search string <see cref="Query.Search"/> to filter settings by settings path.
-        /// This method is called from the filter function <see cref="Main.Query(Query)"/> if the search string <see cref="Query.Search"/> contains the character ">".
+        /// This method is called from the <see cref="Predicate{T}"/> method in <see cref="Main.Query(Query)"/> if the search string <see cref="Query.Search"/> contains the character ">".
         /// </summary>
         /// <param name="found">The WindowsSetting taht should be checked.</param>
-        /// <param name="searchString">The searchString entered by the user <see cref="Query.Search"/>s.</param>
-        internal static bool FilterBySettingsPath(in WindowsSetting found, in string searchString)
+        /// <param name="queryString">The searchString entered by the user <see cref="Query.Search"/>s.</param>
+        internal static bool FilterBySettingsPath(in WindowsSetting found, in string queryString)
         {
-            if (!searchString.Contains('>'))
+            if (!queryString.Contains('>'))
             {
                 return false;
             }
 
-            var pathElements = searchString.Split('>');
+            var queryElements = queryString.Split('>');
 
             // check app name of the WindowsSetting
-            if (!found.Type.StartsWith(pathElements[0]))
+            if (!found.Type.StartsWith(queryElements[0], StringComparison.CurrentCultureIgnoreCase))
             {
                 return false;
             }
@@ -219,22 +219,30 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
             // Check area path of the WindowsSetting.
             // If "pathElements.Length" is higher than one the "searchString" contains at least one area name.
             // Because the area names in pathElements array start at index 1 sometimes <pathElements.Length - 1> or <index - 1> is used.
-            if (pathElements.Length > 1 & !(found.Areas is null))
+            if (!string.IsNullOrEmpty(queryElements[1]))
             {
-                if ((pathElements.Length - 1) > found.Areas.Count())
+                if (found.Areas is null)
+                {
+                    // The user entered at least one area name (pathElements[1] -> pathElements[n]). But the WindowsSetting nas no defined areas.
+                    return false;
+                }
+
+                if ((queryElements.Length - 1) > found.Areas.Count())
                 {
                     // The user entered more area names (pathElements[1] -> pathElements[n]) than defined for this WindowsSetting.
                     return false;
                 }
 
-                for (int i = 1; i <= pathElements.Length; i++)
+                int areaCounter = 0;
+                foreach (var qEelement in queryElements.Skip(1))
                 {
-#pragma warning disable CS8602 // Do not catch exception "Possible null reference". The compiler thinks that Areas could be null which is not possible because of the parental if condition.
-                    if (!found.Areas[i - 1].StartsWith(pathElements[i]))
+                    if (!found.Areas[areaCounter].StartsWith(qEelement, StringComparison.CurrentCultureIgnoreCase))
                     {
                         // The user has entered an area name that not matches.
                         return false;
                     }
+
+                    areaCounter++;
                 }
             }
 

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Helper/ResultHelper.cs
@@ -208,39 +208,36 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings.Helper
                 return false;
             }
 
+            // Init vars
             var queryElements = queryString.Split('>');
 
-            // check app name of the WindowsSetting
-            if (!found.Type.StartsWith(queryElements[0], StringComparison.CurrentCultureIgnoreCase))
+            List<string> settingsPath = new List<string>();
+            settingsPath.Add(found.Type);
+            if (!(found.Areas is null))
             {
-                return false;
+                settingsPath.AddRange(found.Areas);
             }
 
-            // Check area path of the WindowsSetting.
-            if (!string.IsNullOrEmpty(queryElements[1]))
+            // Compare query and settings path
+            for (int i = 0; i < queryElements.Length; i++)
             {
-                if (found.Areas is null)
+                if (string.IsNullOrWhiteSpace(queryElements[i]))
                 {
-                    // The user entered at least one area name (queryElements[1]). But the WindowsSetting has no defined areas.
-                    return false;
+                    // The queryEelent is an WhiteSpace. Nothing to compare.
+                    break;
                 }
 
-                if ((queryElements.Length - 1) > found.Areas.Count())
+                if (i < settingsPath.Count)
                 {
-                    // The user entered more area names (queryElements[1] -> queryElements[n]) than defined for this WindowsSetting.
-                    return false;
-                }
-
-                int areaCounter = 0;
-                foreach (var qElement in queryElements.Skip(1))
-                {
-                    if (!found.Areas[areaCounter].StartsWith(qElement, StringComparison.CurrentCultureIgnoreCase))
+                    if (!settingsPath[i].StartsWith(queryElements[i], StringComparison.CurrentCultureIgnoreCase))
                     {
-                        // The user has entered an area name that not matches.
                         return false;
                     }
-
-                    areaCounter++;
+                }
+                else
+                {
+                    // The user has entered more query parts than existing elements in settings path.
+                    return false;
                 }
             }
 

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Main.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Main.cs
@@ -110,6 +110,12 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings
 
             bool Predicate(WindowsSetting found)
             {
+                if (string.IsNullOrWhiteSpace(query.Search))
+                {
+                    // If no search string is entered skip query comparsion.
+                    return true;
+                }
+
                 if (found.Name.Contains(query.Search, StringComparison.CurrentCultureIgnoreCase))
                 {
                     return true;

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Main.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Main.cs
@@ -112,7 +112,7 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings
             {
                 if (string.IsNullOrWhiteSpace(query.Search))
                 {
-                    // If no search string is entered skip query comparsion.
+                    // If no search string is entered skip query comparison.
                     return true;
                 }
 

--- a/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Main.cs
+++ b/src/modules/launcher/Plugins/Microsoft.PowerToys.Run.Plugin.WindowsSettings/Main.cs
@@ -145,6 +145,12 @@ namespace Microsoft.PowerToys.Run.Plugin.WindowsSettings
                     }
                 }
 
+                // Search by key char '>' for app name and settings path
+                if (query.Search.Contains('>'))
+                {
+                    return ResultHelper.FilterBySettingsPath(found, query.Search);
+                }
+
                 return false;
             }
         }

